ember-headless-form/src/components/-private/control/input.ts

```diff
@@ -28,35 +28,37 @@ export type InputType =
   | 'url'
   | 'week';
 
-export interface HeadlessFormControlInputComponentSignature<VALUE> {
+export interface HeadlessFormControlInputComponentSignature {
   Element: HTMLInputElement;
   Args: {
-    value: VALUE;
+    value: string;
     type?: InputType;
     fieldId: string;
-    setValue: (value: VALUE) => void;
+    setValue: (value: string) => void;
   };
 }
 
-export default class HeadlessFormControlInputComponent<VALUE> extends Component<
-  HeadlessFormControlInputComponentSignature<VALUE>
-> {
-  get type(): InputType {
-    return this.args.type ?? 'text';
-  }
-
-  get valueAsString(): string {
+export default class HeadlessFormControlInputComponent extends Component<HeadlessFormControlInputComponentSignature> {
+  constructor(
+    owner: unknown,
+    args: HeadlessFormControlInputComponentSignature['Args']
+  ) {
     assert(
-      `input can only handle string values, but you passed ${typeof this.args
-        .value}`,
-      typeof this.args.value === 'string'
+      `input component does not support @type="${args.type}" as there is a dedicated component for this. Please use the \`field.${args.type}\` instead!`,
+      args.type === undefined ||
+        // TS would guard us against using an unsupported `InputType`, but for JS consumers we add a dev-only runtime check here
+        (args.type as string) !== 'checkbox'
     );
 
-    return this.args.value;
+    super(owner, args);
+  }
+
+  get type(): InputType {
+    return this.args.type ?? 'text';
   }
 
   @action
   handleInput(e: Event | InputEvent): void {
```

Comment:
Suggestion, nit: Instead of a cast here, can we use a custom type guard?

pseudocode, not tested:

```
function isHTMLInputElement(elem:Element): is HTMLInputElement {
 return elem instanceof HTMLInputElement;
}

// usage
if (isHTMLInputElement(e.target)) { ...do stuff ...}
```

Reply:
Yeah, makes sense. I used `assert('Expected HTMLInputElement', e.target instanceof HTMLInputElement);` now though, as Ember will strip that off a production build, and it does the same type assertion as your `is HTMLInputElement` example.

Reply:
Thanks for the clarification!